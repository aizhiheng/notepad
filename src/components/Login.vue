<template>
  <el-row class="content">
    <el-col :xs="24" :sm="{span: 6,offset: 9}">
      <span class="title">notepad</span>
      <el-row>
        <el-input v-model="account" placeholder="账号" type="text"></el-input>
        <el-input v-model="email" placeholder="邮箱" type="text"></el-input>
        <el-input v-model="password" placeholder="密码" type="password" @keyup.enter.native="loginToDo"></el-input>
        <el-button type="primary" @click="loginToDo">登录</el-button>
        <div><el-button type="primary" @click="register">注册</el-button></div>
      </el-row>
    </el-col>
  </el-row>
</template>

<script>
import md5 from 'md5'

export default {
  data () {
    return {
      account: '',
      email: '',
      password: ''
    }
  },
  methods: {
    loginToDo () {
      let obj = {
        name: this.account,
        password: md5(this.password) // md5加密
      }
      this.$http.post('/auth/user', obj) // 将信息发送给后端
        .then((res) => {
          if (res.data.success) { // 如果成功
            sessionStorage.setItem('demo-token', res.data.token) // 用sessionStorage把token存下来
            this.$message({ // 登录成功，显示提示语
              type: 'success',
              message: '登录成功！'
            })
            this.$router.push('/todolist') // 进入todolist页面，登录成功
          } else {
            this.$message.error(res.data.info) // 登录失败，显示提示语
            sessionStorage.setItem('demo-token', null) // 将token清空
          }
        }, (err) => {
          console.log(err)
          this.$message.error('请求错误！')
          sessionStorage.setItem('demo-token', null) // 将token清空
        })
    },
    register () {
      let params = {
        name: this.account,
        email: this.email,
        password: md5(this.password) // md5加密
      }
      this.$http.post('/auth/createAccount', params)
        .then(response => {
          if (response.data.success) { // 如果成功
            this.$message(response.data.info)
          } else {
            this.$message.error(response.data.info)
          }
        })
        .catch(error => {
          console.log(error)
        })
    }
  }
}
</script>

<style lang="stylus" scoped>
  .el-row.content
    padding 16px
  .title
    font-size 28px
  .el-input
    margin 12px 0
  .el-button
    width 100%
    margin-top 12px
</style>
