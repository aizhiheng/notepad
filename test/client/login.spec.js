import Vue from 'vue'
import elementUI from 'element-ui'
import { mount } from '@vue/test-utils'
import Login from '../../src/components/Login.vue'
import axios from 'axios'

Vue.use(elementUI)

jest.mock('axios', () => ({
  post: jest.fn()
        .mockImplementationOnce(() => Promise.resolve({
          data: {
            success: false,
            info: '用户不存在！'
          }
        }))
        .mockImplementationOnce(() => Promise.resolve({
          data: {
            success: false,
            info: '密码错误！'
          }
        }))
        .mockImplementationOnce(() => Promise.resolve({
          data: {
            success: true,
            token: 'xxx'
          }
        }))
}))

Vue.prototype.$http = axios

let wrapper

const $router = {
  push: jest.fn()
}

beforeEach(() => {
  wrapper = mount(Login, {
    mocks: {
      $router
    }
  })
})

test('Should have two input & one button', () => {
  const inputs = wrapper.findAll('.el-input')
  const loginButton = wrapper.contains('.el-button')
  expect(inputs.length).toBe(2)
  expect(loginButton).toBeTruthy()
})

test('Should have the expected html structure', () => {
  expect(wrapper.element).toMatchSnapshot()
})

test('loginToDo should be called after clicking the button', () => {
  const stub = jest.fn()
  wrapper.setMethods({ loginToDo: stub })
  wrapper.find('.el-button').trigger('click')
  expect(stub).toBeCalled()
})

test('Failed to login if not typing the username & password', async () => {
  const result = await wrapper.vm.loginToDo()
  expect(result.data.success).toBe(false)
  expect(result.data.info).toBe('用户不存在！')
})

test('Failed to login if not typing the correct password', async () => {
  wrapper.setData({
    account: 'molunerfinn',
    password: '1234'
  })
  const result = await wrapper.vm.loginToDo()
  expect(result.data.success).toBe(false)
  expect(result.data.info).toBe('密码错误！')
})

test('Succeeded to login if typing the correct account & password', async () => {
  wrapper.setData({
    account: 'molunerfinn',
    password: '123'
  })
  const result = await wrapper.vm.loginToDo()
  expect(result.data.success).toBe(true)
})
