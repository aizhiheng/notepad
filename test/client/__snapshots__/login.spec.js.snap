// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should have the expected html structure 1`] = `
<div
  class="el-row content"
>
  <div
    class="el-col el-col-24 el-col-xs-24 el-col-sm-6 el-col-sm-offset-9"
  >
    <span
      class="title"
    >
      
     欢迎登录
    
    </span>
     
    <div
      class="el-row"
    >
      <div
        class="el-input"
      >
        <!---->
        <!---->
        <input
          autocomplete="off"
          class="el-input__inner"
          placeholder="账号"
          type="text"
        />
        <!---->
        <!---->
      </div>
       
      <div
        class="el-input"
      >
        <!---->
        <!---->
        <input
          autocomplete="off"
          class="el-input__inner"
          placeholder="密码"
          type="password"
        />
        <!---->
        <!---->
      </div>
       
      <button
        class="el-button el-button--primary"
        type="button"
      >
        <!---->
        <!---->
        <span>
          登录
        </span>
      </button>
    </div>
  </div>
</div>
`;
